tests/blockchain/eth/contracts/unit/test_user_escrow.py

```diff
@@ -446,12 +458,23 @@ def test_upgrading(testerchain, token):
             {'from': user, 'to': user_escrow_contract.address, 'value': 1, 'gas_price': 0})
         testerchain.wait_for_receipt(tx)
 
-    # Only creator can update library
+    # Only creator can update a library
     with pytest.raises((TransactionFailed, ValueError)):
-        tx = linker_contract.functions.upgrade(library_v2.address).transact({'from': user})
+        tx = linker_contract.functions.upgrade(library_v2.address, secret, secret2_hash).transact({'from': user})
         testerchain.wait_for_receipt(tx)
+
+    # Creator must know the secret
+    with pytest.raises((TransactionFailed, ValueError)):
+        tx = linker_contract.functions.upgrade(library_v2.address, secret2, secret2_hash).transact({'from': creator})
+        testerchain.wait_for_receipt(tx)
+
+    # Creator can't use the same secret again because it's insecure
```

Comment:
OK - so secrets are one-time-use in this model (mutated at upgrade-time)?

Reply:
Yes, can't use the same secret in a row